### ISSUE TITLE
Reading is not allowed after reader was completed

### ISSUE THREAD

**Issue by nathanmascitelli** [StackExchange/StackExchange.Redis#2692]
Hi,

We are getting the following exception intermittently in production:
```
at System.IO.Pipelines.ThrowHelper.ThrowInvalidOperationException_NoReadingAllowed()
at System.IO.Pipelines.Pipe.AdvanceReader(SequencePosition& consumed, SequencePosition& examined)
at System.IO.Pipelines.Pipe.DefaultPipeReader.AdvanceTo(SequencePosition consumed, SequencePosition examined)
at StackExchange.Redis.PhysicalConnection.ReadFromPipe() in /_/src/StackExchange.Redis/PhysicalConnection.cs:line 1850
```
When this happens its alway after we have called `IDatabase.StreamAddAsync`.

Looking at PhysicalConnection I'm wondering if there is a race condition between `ReadFromPipe` and something calling `Dispose`? 
![image](https://github.com/StackExchange/StackExchange.Redis/assets/132582942/cbea7db4-b345-4617-8843-45c691e2a0db)

Does that make sense? I'm pretty sure this class is used by multiple threads and so I think the above is posslbe. Looking at the exception handler it seems like we don't even really want to report this exception?
![image](https://github.com/StackExchange/StackExchange.Redis/assets/132582942/bc904fab-f0e7-46c8-8dad-480dbed53fc6)

So to sum up:
- Is this error happening due to the described race condition?
- Do we even want to report this exception up to the caller?

We have seen this happen on multiple OS, dotnet, and package versions:
- Ubuntu 20.04 and 22.04
- Dotnet 7.0.17 and 8.0.3
- StackExchange.Redis 2.7.20 and 2.7.33

Let me know if you need more info.

Thanks!

**Comment by NickCraver** [comment]
It sounds like the connection is being used post-disposal here (at which point we shouldn't be sending any more commands) - is that accurate?

**Comment by nathanmascitelli** [comment]
@NickCraver so the application has one instance of `IConnectionMultiplexer` which is used by multiple threads. At some point one of the threads will dispose the connection (either the app is shutting down or some configruation has changed) yes. When that happens any other threads thats currently trying to use the `IConnectionMultiplexer` gets this error, always when trying to write a value to Redis.

In the screenshot above it looks like we actualy handle this OK when reading but if `isReading` is false then we will always report an exception. Maybe we just need to change that part?

**Comment by nathanmascitelli** [comment]
Any thoughts on my last comment?

**Comment by mukhamedyarov** [comment]
Hello, I am getting the same exceptions. Are there any updates on this?

**Comment by NickCraver** [comment]
Overall here: either do not dispose the multiplexer if you're still using it or stop using it before disposal - this advice is universal to all disposals in .NET. When you dispose an object, that's very intentionally when we cleanup resources and shut things down - if something in in-flight, that's an ill-advised thing to do.

**Comment by nathanmascitelli** [comment]
@NickCraver in general I agree with you, and ultimatly we updated our code to correctly recognize that this exception is equivalent to an `ObjectDisposedException` and act accordingly. However, I still wonder if the catch block should be tweaked so that regardless of if a read is in progress or not we consistanltly report and exception or not if we are in `ReadFromPipe` and the pipe is disposed. It seems odd to me (though maybe I'm just dumb) that sometimes we would report and exception and sometimes not.

**Comment by isacruzramos** [comment]
hi, we are hitting exact same issue of trying to use a disposed multiplexer, are there any guidelines on how to properly identify when the multiplexer was disposed?

**Comment by jianze-mao** [comment]
Hi, we are also getting the same error when trying to read cache. Is there any update on the fix?

**Comment by mgravell** [comment]
Define "fix" in this context. Ultimately, for the reader to be disposed, either the multiplexer was disposed, or a connection died.

Connections do indeed occasionally die - that's a reality of sockets. We try to recover gracefully and reconnect. If the problem is that we did something ungraceful or gave a raw exception: fine, we can probably improve that by presenting a cleaner "oops, something bad happened in transit" exception, but: still ultimately an exception. If we didn't recover at all and got soft-locked: please say, so we can be clear.

If the multiplexer was used after disposing: don't do that. If it was disposed while something was in flight, and the in flight operation caused a viscous side effect: please be explicit.

So: more detail please. Rather than just saying "same error": what *exactly* happened, and what *exactly* was your doing at the time? Details matter, sorry.

**Comment by fstugren** [comment]
I see the same error in our services, and the connection multiplexer is not disposed. The service uses a shared Redis cache client with a connection multiplexer that lives for the duration of the application. When those "_Reading is not allowed after reader was completed_" exceptions happen the application pod is not shutting down; the error is generally isolated and eventually recovers. Sometimes there's a burst of errors like this one, sometimes just one. Eventually, it appears that the Redis client successfully reconnects, and operations resume.

What is the recommendation on the client side? We could catch this specific exception and ignore it, but the problem is that each time we run a "Get" operation that fails with this error that it takes 3-5 seconds. Is there anything else?

**Comment by mgravell** [comment]
Thanks for the report. StackExchange.Redis **3.0** ships a substantial rewrite of the low-level IO core — a new RESPite-based stack that replaces `Pipelines.Sockets.Unofficial` and the old backlog/write loop — and we're confident this issue is addressed by that work.

Closing as resolved in 3.0. If you can still reproduce on 3.0.x, please comment and we'll reopen and investigate with fresh diagnostics.
